### PR TITLE
Add support for uid:gid options

### LINKED ISSUE #18: change permissions of newly created volumes to uid:gid
Hey there,

I was wondering if it was possible in concept to automatically chown a newly created volumes to a specific user. While generally being awesome, we have to change containers to run as root sometimes just to automatically be able to create folder structures.

If you think it is generally possible, I'd be happy to send a PR your way :)

### PR DESCRIPTION
**Pull request by s4ke**
closes #18

### CODE REVIEW

**Comment by s4ke** [comment]
Have not 100% tested the driver yet, but creating the PR for review already :)

File: Dockerfile [hunk 1]
```diff
@@ -6,12 +6,8 @@ RUN apk add --update git
 
 WORKDIR /plugin
 
-# warm up go mod cache
-COPY go.mod go.sum ./
-RUN go mod download
-
 COPY . /plugin
-RUN go build -v
+RUN go build -v -mod vendor
```
Comment (costela) [inline Dockerfile]: what's the rationale behind vendoring? This seems like an unrelated change?

File: driver.go [hunk 2]
```diff
@@ -98,6 +98,27 @@ func (hd *hetznerDriver) Create(req *volume.CreateRequest) error {
 		}
 	}
 
+	uid := getOption("uid", req.Options)
+	gid := getOption("gid", req.Options)
+	if uid != "0" || gid != "" {
```
Comment (costela) [inline driver.go]: the default value for `gid` above is "0", right? shouldn't we change this to match it?

File: driver.go [hunk 3]
```diff
@@ -98,6 +98,27 @@ func (hd *hetznerDriver) Create(req *volume.CreateRequest) error {
...
+		// string to int
+		uintParsed, err := strconv.Atoi(uid)
+		if err != nil {
+			logrus.Warnf("failed to parse uid option value as uint32: %s", uid)
+			uintParsed = 0
+		}
+		gidParsed, err := strconv.Atoi(gid)
+		if err != nil {
+			logrus.Warnf("failed to parse gid option value as uint32: %s", gid)
```
Comment (costela) [inline driver.go]: nit: `strconv.Atoi` returns `int`, not `uint32` (also: the exact go type is probably irrelevant for the user seeing the error; I think the important part is the "integer")

File: driver.go [hunk 4]
```diff
@@ -98,6 +98,27 @@ func (hd *hetznerDriver) Create(req *volume.CreateRequest) error {
...
+			gidParsed = 0
+		}
```
Comment (costela) [inline driver.go:112]: I think we have an interesting error scenario here: if the user sets uid/gid to "foo", they effectively cause a `chown root:root`. This may not be the best option. Shouldn't we just skip chowning completely on parse errors instead?

File: os.go [hunk 5]
```diff
@@ -34,3 +35,80 @@ func mkfs(dev, fstype string) error {
...
+		"/bin/mount",
+		mountArgs...,
+	)
+	cmd.Stderr = &stderr
+	if err := cmd.Run(); err != nil {
+		logrus.Errorf("mount stderr: %s", stderr.String())
+		return err
+	}
+	return nil
+}
```
Comment (costela) [inline os.go]: can we maybe use something like [github.com/moby/sys.Mount](https://pkg.go.dev/github.com/moby/sys@v0.0.0-20200315194828-07b9aaad3088/mount#Mount) and Umount here? I'm already a bit wary about using `exec.Command` for `mkfs` :sweat:

File: os.go [hunk 6]
```diff
@@ -34,3 +35,80 @@ func mkfs(dev, fstype string) error {
...
+		uidgid,
+		dir,
+	)
+	cmd.Stderr = &stderr
+	if err := cmd.Run(); err != nil {
+		logrus.Errorf("chown stderr: %s", stderr.String())
+		return err
+	}
+	return nil
+}
```
Comment (costela) [inline os.go]: I _think_ this can be replaced with `fs.WalkDir` + `os.Chown`? This would get rid of yet another `exec.Command` instance :+1:

File: driver.go [hunk 7]
```diff
@@ -98,6 +98,27 @@ func (hd *hetznerDriver) Create(req *volume.CreateRequest) error {
...
+
+		err = setPermissions(resp.Volume.LinuxDevice, getOption("fstype", req.Options), uint32(uintParsed), uint32(gidParsed))
+		if err != nil {
```
Comment (costela) [inline driver.go]: nit (style):
```suggestion
		if err := setPermissions(resp.Volume.LinuxDevice, getOption("fstype", req.Options), uint32(uintParsed), uint32(gidParsed)); err != nil {
```

**Review by costela — CHANGES_REQUESTED** [review]
Thanks a lot for the contribution! :rocket: 
Just have a few suggestions/questions.

File: driver.go (hunk 2 above)
Comment (s4ke) [reply driver.go]: ouch. good catch

File: Dockerfile (hunk 1 above)
Comment (s4ke) [reply Dockerfile]: Tbh, I am a bit new to Go. I am mostly copying from what I know from moby/swarmkit and moby/moby where I also added some PRs.

What is the reason for not vendoring vs vendoring? If you are not against vendoring, but mostly are concerned with this not being part of THIS pr, I am happy to remove it from the PR

File: driver.go (hunk 4 above)
Comment (s4ke) [reply driver.go:112]: Ok, this is a good point. I will adapt the PR accordingly

File: os.go (hunk 5 above)
Comment (s4ke) [reply os.go]: sure, I will check what I can do. I mostly copied the style, thanks for the suggestion

File: os.go (hunk 6 above)
Comment (s4ke) [reply os.go]: We can try this.

Tbh, I wonder whether we even need the fs.WalkDir then, as we are mounting an empty volume that at most contains a lost+found folder which I honestly think should be irrelevant anyways.

File: Dockerfile (hunk 1 above)
Comment (costela) [reply Dockerfile]: I try to avoid vendoring unless necessary: it kinda breaks some parts of the go trust-model. When using modules from the internet, you can kinda trust that everybody gets the same code for the dependencies (via mod proxy + sumdb, which can be audited). When using vendoring you - as an application consumer - can't be sure that the code inside `/vendor` is actually the same as the module name would indicate. You have to check.

Some people use vendoring as a way to speed up CI pipelines (IMHO not worth the trade-off) or to keep non-upstreamed changes local to the project (which is less work than upstreaming every change, but IMHO a bit short-sighted).

File: os.go (hunk 6 above)
Comment (costela) [reply os.go]: ah yes, good point. It might even be a good place to _check_ if the folder is empty and warn if it isn't?

**Comment by s4ke** [comment]
I think all of the issues should be adressed now, but I think we should do some testing on Hetzner VMs before we merge this.

**Comment by costela** [comment]
Sounds reasonable. I'm not sure when I'll find the time though :unamused: 
We could do an RC release for that.

File: driver.go [hunk 8]
```diff
@@ -98,6 +98,24 @@ func (hd *hetznerDriver) Create(req *volume.CreateRequest) error {
...
+			return errors.Wrapf(err, "failed to parse uid option value as integer: %s", gid)
+		}
+		gidParsed, err := strconv.Atoi(gid)
+		if err != nil {
+			return errors.Wrapf(err, "failed to parse gid option value as integer: %s", gid)
+		}
+
+		if err := setPermissions(resp.Volume.LinuxDevice, getOption("fstype", req.Options), uintParsed, gidParsed, ""); err != nil {
+			return errors.Wrapf(err, "could not chown '%s' to '%s:%s'", resp.Volume.LinuxDevice, uid, gid)
```
Comment (costela) [inline driver.go]: nit: 
```suggestion
			return errors.Wrapf(err, "could not chown %q to '%s:%s'", resp.Volume.LinuxDevice, uid, gid)
```

File: os.go [hunk 9]
```diff
@@ -34,3 +36,54 @@ func mkfs(dev, fstype string) error {
 	}
 	return nil
 }
+
+func chownIfEmpty(dir string, uid int, gid int) error {
+	files, err := ioutil.ReadDir(dir)
```
Comment (costela) [inline os.go]: sorry for annoying further, but `ioutil` functions are deprecated. Can we maybe use `os.ReadDir` instead? :sweat_smile:

File: os.go [hunk 10]
```diff
@@ -34,3 +36,54 @@ func mkfs(dev, fstype string) error {
...
+		logrus.Errorf("chown error: %s", err)
+		return err
+	}
+	return nil
+}
+
+func setPermissions(dev, fstype string, uid int, gid int, mountOptions string) error {
+	tmpDir, err := os.MkdirTemp(os.TempDir(), "mnt-*")
+	if err != nil {
+		return fmt.Errorf("failed creating temp dir for setting permissions")
```
Comment (costela) [inline os.go]:
```suggestion
		return fmt.Errorf("failed creating temp dir for setting permissions: %w", err)
```

File: os_test.go [hunk 11]
```diff
@@ -0,0 +1,142 @@
...
+	"os"
+	"syscall"
+	"testing"
+
+	"github.com/docker/docker/pkg/mount"
+	"github.com/sirupsen/logrus"
+)
+
+func Test_setPermissions(t *testing.T) {
+	t.Run("test", func(t *testing.T) {
```
Comment (costela) [inline os_test.go]: if we only have a single test case, we can skip `t.Run`, right?

**Review by costela — CHANGES_REQUESTED** [review]
sorry, just noticed a couple of small things

**Comment by s4ke** [comment]
no need to be sorry, this is a free GoLang course for me 😆 

I will adapt the code and come back with another attempt!

**Comment by s4ke** [comment]
I will do some tests on an actual Swarm over the weekend and Update with the results here

**Comment by s4ke** [comment]
With my custom built image at ancieque/docker-volume-hetzner:dev everything worked :)

grafik

However, only the uid has changed for some reason, but that might be caused by me mistyping and writing "pid" instead of "gid" :D

**Comment by s4ke** [comment]
Update: with proper gid, everything works:

grafik

This looks mergeable to me now!

**Comment by s4ke** [comment]
@costela Please let me know if you prefer a rebased PR regarding git commit history :)

**Review by costela — APPROVED** [review]
All good! Thanks for the contribution! :tada:

**Comment by s4ke** [comment]
Awesome. Can we get a new release for this? Or do you prefer if we build this ourselves?

**Comment by costela** [comment]
I'm doing a bit of cleanup and intend on cutting a release in the next few days. With a bit of luck maybe tonight. :crossed_fingers:

**Comment by costela** [comment]
new release is out :rocket:
feel free to give it a spin!

**Comment by s4ke** [comment]
Tyvm. Feels good to have a release after a while again :).

Thanks for the patience with the PR and for the free code kata :rocket:
